frontend/public/components/utils/storage-class-dropdown.tsx

```diff
@@ -0,0 +1,196 @@
+  };
+
+  componentWillMount() {
+    this.componentWillReceiveProps(this.props);
+  }
+
+  componentWillReceiveProps(nextProps) {
+    const { loaded, loadError, resources } = nextProps;
+
+    if (loadError) {
+      this.setState({
+        title: (
+          <div className="cos-error-title">Error Loading {nextProps.desc}</div>
+        ),
+      });
+      return;
+    }
+    if (!loaded) {
+      return;
+    }
+
+    const state = {
+      items: {},
+      title: {},
+      defaultClass: '',
+    };
+    const unorderedItems = {};
+    const noStorageClass = 'No storage class';
+    _.each(resources, ({ data }, kindLabel) => {
+      _.reduce(
+        data,
+        (acc, resource) => {
+          acc[resource.metadata.name] = {
+            kindLabel,
+            name: resource.metadata.name,
+            description: _.get(resource, 'metadata.annotations.description', ''),
+            default: _.get(resource, ['metadata', 'annotations', 'storageclass.kubernetes.io/is-default-class']) === 'true',
+            accessMode:  _.get(resource, ['metadata', 'annotations', 'storage.alpha.openshift.io/access-mode'], ''),
+            provisioner: resource.provisioner,
+            type: _.get(resource, 'parameters.type', ''),
+            zone: _.get(resource, 'parameters.zone', ''),
+          };
+          return acc;
+        },
+        unorderedItems
+      );
+    });
+
+    // Determine if there is a default storage class
+    state.defaultClass = _.findKey(unorderedItems, 'default');
+    const { selectedKey } = this.state;
+    if (!state.defaultClass) {
+      // Add No Storage Class option if there is not a default storage class
+      unorderedItems[''] = { kindLabel: '', name: noStorageClass };
+    }
+
+    if (!this.props.loaded || !selectedKey ) {
+      state.title = <span className="text-muted">Select storage class</span>;
+    }
+
+    const selectedItem = unorderedItems[selectedKey || state.defaultClass];
+    if (selectedItem) {
+      state.title = this.getTitle(selectedItem);
+    }
+
+    Object.keys(unorderedItems).sort().forEach((key) => {
+      state.items[key] = unorderedItems[key];
+    });
+    this.setState(state);
+  }
+
+  componentDidUpdate() {
+    const {defaultClass, selectedKey } = this.state;
+    if (!selectedKey && defaultClass) {
+      this.onChange(defaultClass);
+    }
+  }
+
+  shouldComponentUpdate(nextProps, nextState) {
+    return !_.isEqual(this.state, nextState);
+  }
+
+  autocompleteFilter = (text, item) => fuzzy(text, item.props.name);
+
+  getTitle = storageClass => {
+    return storageClass.kindLabel ? <ResourceName kind="StorageClass" name={storageClass.name} /> : <span >{storageClass.name}</span>;
+  }
+
+  onChange = key => {
+    const storageClass = _.get(this.state, ['items', key], {});
+    this.setState({
+      selectedKey: key,
+      title: this.getTitle(storageClass),
+    });
+    this.props.onChange(_.find(this.props.resources.StorageClass.data, item => {
```

Comment:
Why not keep the original storage class object as a property on item? Then we don't need to find it again.